pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -370,10 +466,77 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 	}
 	var err error
 	for k, v := range configs {
-		resources, err = addConfigMapItem(master, getConfigName(master, k), labels, v, resources)
+		resources, err = addConfigMapItem(master, getResourceName(master, k), labels, v, resources)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If upgrade has failed, do not set imageToUse, just return
+	if master.Status.Meta.IsConditionTrue(upgradeFailed) {
+		// User has reset the CR after failure. Remove upgrade failed status.
+		if master.Spec.Image == master.Status.ImageToUse {
+			master.Status.ClearCondition(upgradeFailed, upgradeResetMessage, upgradeResetMessage)
+		}
+		return resources, nil
+	}
+
+	// Early return if image does not need updating. Update imageToUse since we are in a non
+	// failure case.
+	if master.Status.ImageToUse == "" || master.Status.ImageToUse == master.Spec.Image {
+		master.Status.ImageToUse = master.Spec.Image
+		master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
```

Comment:
The intention of having two docker images for the backend and the UI is to allow them upgrade separately. It may not be an issue now by assuming they get upgraded together, but it's good to note that it may be different in future.